Add phrase search (#1021)
meilisearch/documentation: reference/features/search_parameters.md

Placeholder search returns the documents that best match other search parameters, sorted according to that index's [ranking rules](/reference/features/settings.md#ranking-rules). This feature also supports [faceting](/reference/features/faceted_search.md) and [filtering](/reference/features/filtering.md)

### Phrase search

You can enclose search terms in double quotes (`"`), in which case MeiliSearch will only return documents containing all terms in the order they were given. This is called a [phrase search](/learn/what_is_meilisearch/features.md#phrase-search).

> Phrase searches are case-insensitive and ignore soft separators such as `-`, `,`, and `:`. You can mix them with regular queries: MeiliSearch will first fetch all documents with exact matches and [then proceed with its default behavior](/learn/core_concepts/relevancy.md).

Reviewer: ```suggestion
Phrase searches ignore soft separators such as `-`, `,`, and `:`. You can mix them with regular queries: MeiliSearch will first fetch all documents with exact matches and [then proceed with its default behavior](/learn/core_concepts/relevancy.md).
```

Author: Thanks for the input, @curquiza! Does this mean we are deviating from the spec? 

For reference: [phrase query spec, usage heading](https://github.com/meilisearch/specifications/blob/main/text/0043-phrase-query.md#phrase-query-usage)

Reviewer: Indeed @ManyTheFish can you confirm the mistake?
FYI @gmourier

Reviewer 2: Phrase searches are case-insensitive.

Reviewer: Sorry! I confused case-sensitive and case-insensitive!!!! My bad everyone! 😂 I definitely need holidays!

Reviewer 3: Should we link to the full list of soft separators?

Author: I thought about that, but couldn't find an in-house resource. Do you have a good external resource we could link to?

Reviewer 3: We have a documentation page in the "Under the hood" section that describes our soft and hard separators https://docs.meilisearch.com/reference/under_the_hood/datatypes.html but it does not describe the full list.

Here the source code https://github.com/meilisearch/tokenizer/blob/0db4b8fca263bfc89cb4c4453877a95e63806a73/src/detection.rs#L49

@ManyTheFish Please tell me if I'm wrong about the file 😇. Also, I didn't see the `—` char we identified some time ago. We had added an issue but it seems to me that there was a mix-up do you recall?

Author: @gmourier thanks for the link! 

Is there a reason why the datatypes page doesn't list all separators? Do we use different lists of soft/hard separators depending on what we're doing?

If not, I'll open an issue and update it.

> 

Reviewer 2: We spoke about soft separators, but hard separators are handled differently. When we have a Hard separator in a phrase search, we split the phrase into 2 different phrases.
`hello "world! Nice to see you"` would be considered as `hello "world" "Nice to see you"`.
Because in meilisearch, a Hard separator is considered as a split of 2 unrelated contexts.

Author: Thanks for that, @ManyTheFish — I'll add an extra sentence about hard separators.

Placeholder search returns the documents that best match other search parameters, sorted according to that index's [ranking rules](/reference/features/settings.md#ranking-rules). This feature also supports [faceting](/reference/features/faceted_search.md) and [filtering](/reference/features/filtering.md)

### Phrase search

> You can enclose search terms in double quotes (`"`), in which case MeiliSearch will only return documents containing all terms in the order they were given. This is called a [phrase search](/learn/what_is_meilisearch/features.md#phrase-search).

Reviewer 4: ```suggestion
If you enclose search terms in double quotes (`"`), MeiliSearch will only return documents containing those terms in the order they were given. This is called a **phrase search**.
```
Feels weird to link to a less detailed section on the same subject :)

> Phrase searches are case-insensitive and ignore [soft separators such as `-`, `,`, and `:`](/reference/under_the_hood/datatypes.md). Hard separators within a phrase search effectively split the query into different phrase searches: `"Octavia.Butler"` will return the same results as `"Octavia" "Butler"`.

Reviewer 4: ```suggestion
Phrase searches are case-insensitive and ignore [soft separators such as `-`, `,`, and `:`](/reference/under_the_hood/datatypes.md). Using a hard separator within a phrase search effectively splits it into multiple separate phrase searches: `"Octavia.Butler"` will return the same results as `"Octavia" "Butler"`.
```

> You can mix phrase searches with regular queries: MeiliSearch will first fetch all documents with exact matches and [then proceed with its default behavior](/learn/core_concepts/relevancy.md).

Reviewer 4: ```suggestion
You can combine phrase search and normal queries in a single search request. In this case, MeiliSearch will first fetch all documents with exact matches to the given phrase(s), and [then proceed with its default behavior](/learn/core_concepts/relevancy.md).
```

> You can combine phrase search and normal queries in a single search request. In this case, MeiliSearch will first fetch all documents with exact matches to the given phrase(s), and [then proceed with its default behavior](/learn/core_concepts/relevancy.md).

Reviewer 4: I'm having second thoughts about this last sentence, not sure it adds anything. What do you think of getting rid of it?

```suggestion
You can combine phrase search and normal queries in a single search request.
```

Author: I'm leaning towards leaving it because this explains exactly why the example (`"african american" poet`) works.

Reviewer 4: Hmm, I don't completely understand but it's probably just the sleep deprivation talking.

Author: `"African American" poet` works well because it matches `African American` strictly , but uses the regular MeiliSearch process for `poet`, meaning it will match also match words like `poem` (possibly also `poetic` and `poetry`? Not 100% sure on those two, to be honest).

Reviewer 2: Hey @guimachiavelli! In your example `poet` has less than 5 characters, meaning that we don't allow a typo on this word 😅 but if the word would be longer we would allow 1 typo on this word.

Meanwhile, because `poet` is not in the `".."`, meilisearch will return documents where `poet` isn't near `African American`.

Author: @ManyTheFish I stand corrected! Thanks for that!
